### PR TITLE
Fix raft state never being sent

### PR DESCRIPTION
**Pull request by Serpentian**
Closes https://github.com/tarantool/tarantool/issues/10089
Closes https://github.com/tarantool/tarantool/issues/10727

### CODE REVIEW

File: changelogs/unreleased/gh-10089-raft-meta-join.md
Comment (andreyaksenov) [inline changelogs/unreleased/gh-10089-raft-meta-join.md:4]:
```suggestion
* Fixed a bug when the Raft state wasn't sent during the `META_JOIN` stage, which could
  lead to split-brain (gh-10089).
```

**Review by andreyaksenov — APPROVED** [review]

File: test/replication-luatest/join_test.lua
Comment (Gerold103) [inline test/replication-luatest/join_test.lua:69]: `is` -> `was`. To make the emphasis that now it is not so, and that is good.

**Review by Gerold103 — APPROVED** [review]
Good work on spotting and swiftly fixing the bug 💪.

File: src/box/xrow.c
Comment (Serpentian) [reply src/box/xrow.c:1537]: > I would suggest to have just one xrow_decode_raft which can work with any group (which won't even check it). That would be a bit simpler.

I intentionally did that in order to fortify the code execution. I suppose, this way the code will be more strict and we'll think twice, whether the row, we're reading, is local or remote, since RAFT can be both. This way we won't repeat this mistake in the future.

I'd rather leave as it is. Let's see what @sergepetrenko think here

Comment (sergepetrenko) [reply src/box/xrow.c:1537]: I think you could put the `group_id` into `raft_request`/`raft_msg` structs and make a single version of `xrow_encode/decode_raft` which handles the group id.

That's because we already have `raft_checkpoin_local` and `raft_checkpoint_remote`, which may set group id as well.

File: changelogs/unreleased/gh-10089-raft-meta-join.md
Comment (Serpentian) [reply changelogs/unreleased/gh-10089-raft-meta-join.md]: Fixed

File: test/replication-luatest/join_test.lua
Comment (Serpentian) [reply test/replication-luatest/join_test.lua]: Fixed

File: src/box/xrow.c
Comment (Serpentian) [reply src/box/xrow.c]: > I think you could put the group_id into raft_request/raft_msg structs and make a single version of xrow_encode/decode_raft which handles the group id.

Great idea! Done that. So much cleaner now

**Comment by Serpentian** [comment]
@Gerold103, since the patch changed too majorly, I'm asking you to review it one more time)

File: src/lib/raft/raft.h
Comment (Gerold103) [inline src/lib/raft/raft.h:132]: Wait wait, lib/raft shouldn't care about this. It only produces messages and checkpoints. Group ID is a thing made up for our replication, unrelated to the basic Raft. Lets move this all into box/raft. It is built on top of lib/raft and can deal with all the Tarantool replication specifics.

I know it seems convenient now, but lib/raft doesn't use this flag, only sets it. Which means we leak the logic from an upper module into the lower module, and that would hit us eventually if more things come into lib/raft like this, in terms of code complexity and dependencies.

**Review by Gerold103 — COMMENTED** [review]
Thanks for the fixes, it does look cleaner indeed! However we introduce into lib/raft some alien logic now 😰.

File: src/lib/raft/raft.h
Comment (Serpentian) [reply src/lib/raft/raft.h]: Thank you for noticing! Didn't think about that. Now it's even simpler, but I had to drop the assertions from the `raft_process_recovery`, since now, when raft state is actually received, the message has `state` and may have `vclock`.

**Review by Gerold103 — APPROVED** [review]
Thanks for addressing all my comments 🙏!

File: src/box/xrow.c
Comment (locker) [inline src/box/xrow.c:1481]: After this change, a replica running an older binary won't be able to subscribe to a master running a patched binary. This would break even patch-level upgrades.

File: src/box/relay.cc
Comment (Serpentian) [inline src/box/relay.cc:1122]: This is the least crutchy solution I managed to find. The solution for initial https://github.com/tarantool/tarantool/issues/10089 is easy, we can just patch `relay_send_initial_join_row` to make it send raft requests, but I don't want to send `GROUP_LOCAL` rows at all, this seems incorrect ideologically.

It would be perfect, if sending raft state from `box_process_subscribe` could be done via `relay_send`, but I didn't manage to make it work, seems like it's there for the reason.

**Comment by Serpentian** [comment]
@Gerold103, since @locker found a major bug in the prev iteration, it was rewritten again, I need your review)

File: src/box/relay.cc
Comment (locker) [reply src/box/relay.cc:1122]: The label `2.11` to #10089 suggests that it should be backported to 2.11 so we need to take it into account when checking `version_id`. Also, since the version check is used in two places, I'd put it in a helper function.

**Review by locker — APPROVED** [review]

File: test/replication-luatest/gh_10727_no_local_rows_test.lua
Comment (Gerold103) [inline test/replication-luatest/gh_10727_no_local_rows_test.lua:24]: Nit: lets avoid pronouns in the comments like "I", "me", "my", etc. The code authorship is going to be not quite easily deductible after it is merged into master.

File: src/box/relay.cc
Comment (Gerold103) [inline src/box/relay.cc:1129]: Hm. Having it in such a generic place feels quite crutchy, like you said. Perhaps extract Raft row sending into a new function called `relay_send_raft()` containing this check and invoke it explicitly in `relay_send_tx()` (when request type is Raft) and in `relay_raft_msg_push()` unconditionally?

Then we could at least isolate the hack.

File: test/replication-luatest/gh_10727_no_local_rows_test.lua
Comment (Gerold103) [inline test/replication-luatest/gh_10727_no_local_rows_test.lua:25]: You could perhaps make this function a closure and wakeup the main test fiber when the condition is found. The test case would then terminate the proxy and do all the proper shutdown like a regular test.

**Review by Gerold103 — COMMENTED** [review]
Thanks for fixing this, and shame on me that I didn't notice the compatibility problem 😰.

File: test/replication-luatest/gh_10727_no_local_rows_test.lua
Comment (Serpentian) [reply test/replication-luatest/gh_10727_no_local_rows_test.lua]: Or we can use a global variable) Even simpler

File: src/box/relay.cc
Comment (Serpentian) [reply src/box/relay.cc]: WDYT of the current implementation? `relay_send_tx` never sends raft, all local rows are filtered out from it. You meant `relay_send_initial_join_row`, I suppose

File: test/replication-luatest/gh_10727_no_local_rows_test.lua
Comment (Serpentian) [reply test/replication-luatest/gh_10727_no_local_rows_test.lua]: The comment doesn't exist anymore. I got you, for the future

File: src/box/relay.cc
Comment (Gerold103) [reply src/box/relay.cc]: Aah, yep, then in `relay_send_initial_join_row()`. The rest of the implementation is fine! I can't think of any other way.

Comment (Serpentian) [reply src/box/relay.cc]: @Gerold103, please, approve to PR, if we're good to go)

File: src/box/relay.cc
Comment (Gerold103) [inline src/box/relay.cc:1159]: Please, lets turn the `if (is raft request)` inside `relay_send_raft()` into an assertion. Otherwise it looks quite strange. When a request is not Raft, the function called "send raft" just ignores it. As a result, either the caller already knows it is a Raft request and uses this function, or has to check the type before calling the function. Either way the `if` inside of the function never fails. Which means it is an assertion. No?

**Review by Gerold103 — COMMENTED** [review]
Sorry, just one last nit.

File: src/box/relay.cc
Comment (Serpentian) [reply src/box/relay.cc:1157]: Agree, fixed

**Review by Gerold103 — APPROVED** [review]
Prime top 💪!

File: src/box/xrow.c
Comment (sergepetrenko) [reply src/box/xrow.c:1481]: Resolving since the issue is addressed.

File: src/box/relay.cc
Comment (sergepetrenko) [reply src/box/relay.cc]: Resolving since the issue is addressed

**Review by sergepetrenko — APPROVED** [review]
Thanks for the patch!
LGTM.

**Pull request merged**
